File: ShellGUI/lib/core/invoker.dart
// ignore_for_file: unused_import, library_prefixes

import '/common.dart';
import 'dart:ffi' as ffi;
import 'package:ffi/ffi.dart' as ffi;
import '/core/interface.dart' as Interface;
import '/core/converter.dart';
import '/core/library.dart';

// ----------------

class Invoker {

  static List<String> Function(List<String>)? _callbackTarget;

  static ffi.Pointer<Interface.String>?       _callbackExceptionHandler;

  static ffi.Pointer<Interface.StringList>?   _callbackResultHandler;

  static
  ffi.Pointer<Interface.String>
  _callbackProxy(
    ffi.Pointer<ffi.Pointer<Interface.StringList>> argument,
    ffi.Pointer<ffi.Pointer<Interface.StringList>> result,
  ) {
    Converter.destructString(_callbackExceptionHandler!.ref);
    Converter.destructStringList(_callbackResultHandler!.ref);
    result.value = _callbackResultHandler!;
    try {
      var argumentValue = Converter.parseStringList(argument.value.ref);
      var resultValue = _callbackTarget!(argumentValue);
      Converter.constructStringList(_callbackResultHandler!.ref, resultValue);
      return ffi.nullptr;
    } catch (e) {
      var exceptionValue = e.toString();
      Converter.constructString(_callbackExceptionHandler!.ref, exceptionValue);
      return _callbackExceptionHandler!;
    }
  }

  // ----------------

  static
  Integer
  version(
    Library library,
  ) {
    var numberPointer = ffi.Pointer<ffi.Pointer<Interface.Size>>.fromAddress(0);
    {
      numberPointer = ffi.calloc.call<ffi.Pointer<Interface.Size>>();
    }
    {
    }
    var exceptionPointer = library.version(numberPointer);
    {
    }
    var number = Converter.parseSize(numberPointer.value.ref);
    {
      ffi.calloc.free(numberPointer);
    }
    if (exceptionPointer != ffi.nullptr) {
      var exception = Converter.parseString(exceptionPointer.ref);
      throw exception;
    }
    return number;
  }

  static
  String
  execute(
    Library                             library,
    List<String> Function(List<String>) callback,
    String                              script,
    List<String>                        argument,
  ) {
    {
      _callbackTarget = callback;
      _callbackExceptionHandler = ffi.calloc.call<Interface.String>();
      _callbackResultHandler = ffi.calloc.call<Interface.StringList>();
      Converter.constructStringList(_callbackResultHandler!.ref, []);
      Converter.constructString(_callbackExceptionHandler!.ref, '');
    }
    var callbackPointer = ffi.Pointer<ffi.Pointer<Interface.Callback>>.fromAddress(0);
    var scriptPointer = ffi.Pointer<ffi.Pointer<Interface.String>>.fromAddress(0);
    var argumentPointer = ffi.Pointer<ffi.Pointer<Interface.StringList>>.fromAddress(0);
    var resultPointer = ffi.Pointer<ffi.Pointer<Interface.String>>.fromAddress(0);
    {
      callbackPointer = ffi.calloc.call<ffi.Pointer<Interface.Callback>>();
      scriptPointer = ffi.calloc.call<ffi.Pointer<Interface.String>>();
      argumentPointer = ffi.calloc.call<ffi.Pointer<Interface.StringList>>();
      resultPointer = ffi.calloc.call<ffi.Pointer<Interface.String>>();
    }
    {
      callbackPointer.value = ffi.calloc.call<Interface.Callback>();
      scriptPointer.value = ffi.calloc.call<Interface.String>();
      argumentPointer.value = ffi.calloc.call<Interface.StringList>();
      Converter.constructCallback(callbackPointer.value.ref, ffi.Pointer.fromFunction(_callbackProxy));
      Converter.constructString(scriptPointer.value.ref, script);
      Converter.constructStringList(argumentPointer.value.ref, argument);
    }
    var exceptionPointer = library.execute(callbackPointer, scriptPointer, argumentPointer, resultPointer);
    {
      Converter.destructCallback(callbackPointer.value.ref);
      Converter.destructString(scriptPointer.value.ref);
      Converter.destructStringList(argumentPointer.value.ref);
      ffi.calloc.free(callbackPointer.value);
      ffi.calloc.free(scriptPointer.value);
      ffi.calloc.free(argumentPointer.value);
    }
    var result = Converter.parseString(resultPointer.value.ref);
    {
      ffi.calloc.free(callbackPointer);
      ffi.calloc.free(scriptPointer);
      ffi.calloc.free(argumentPointer);
      ffi.calloc.free(resultPointer);
    }
    {
      Converter.destructStringList(_callbackResultHandler!.ref);
      Converter.destructString(_callbackExceptionHandler!.ref);
      ffi.calloc.free(_callbackExceptionHandler!);
      ffi.calloc.free(_callbackResultHandler!);
      _callbackTarget = null;
    }
    if (exceptionPointer != ffi.nullptr) {
      var exception = Converter.parseString(exceptionPointer.ref);
      throw exception;
    }
    return result;
  }

  static
  Void
  prepare(
    Library library,
  ) {
    {
    }
    {
    }
    var exceptionPointer = library.prepare();
    {
    }
    {
    }
    if (exceptionPointer != ffi.nullptr) {
      var exception = Converter.parseString(exceptionPointer.ref);
      throw exception;
    }
    return;
  }

}

File: Helper/Core/Invoker.cs
#pragma warning disable 0,
// ReSharper disable InconsistentNaming VariableHidesOuterVariable AccessToModifiedClosure

using Helper;

namespace Helper.Core {

	public static unsafe class Invoker {

		#region function

		public static IntegerU64 Version (
			Library library
		) {
			var numberPointer = (Interface.Size*)null;
			{
				fixed (Interface.Size* pointer = new Interface.Size[1]) {
					*pointer = Converter.CreateSize(0);
					numberPointer = pointer;
				}
			}
			var exceptionPointer = library.Version(&numberPointer);
			var number = Converter.ParseSize(ref *numberPointer);
			if (exceptionPointer != null) {
				var exception = Converter.ParseString(ref *exceptionPointer);
				throw new Exception(exception);
			}
			return number;
		}

		public static String Execute (
			Library                          library,
			Func<List<String>, List<String>> callback,
			String                           script,
			List<String>                     argument
		) {
			var callbackExceptionHandler = (Interface.String*)null;
			var callbackResultHandler = (Interface.StringList*)null;
			var callbackProxy = Interface.String* (
				Interface.StringList** argument,
				Interface.StringList** result
			) => {
				*result = callbackResultHandler;
				try {
					var argumentValue = Converter.ParseStringList(ref **argument);
					var resultValue = callback(argumentValue);
					*callbackResultHandler = Converter.CreateStringList(resultValue);
					return null;
				} catch (Exception e) {
					var exceptionValue = e.ToString();
					*callbackExceptionHandler = Converter.CreateString(exceptionValue);
					return callbackExceptionHandler;
				}
			};
			fixed (Interface.String* pointer = new Interface.String[1]) {
				callbackExceptionHandler = pointer;
				*callbackExceptionHandler = Converter.CreateString("");
			}
			fixed (Interface.StringList* pointer = new Interface.StringList[1]) {
				callbackResultHandler = pointer;
				*callbackResultHandler = Converter.CreateStringList(new List<String>());
			}
			var callbackPointer = (Interface.Callback*)null;
			var scriptPointer = (Interface.String*)null;
			var argumentPointer = (Interface.StringList*)null;
			var resultPointer = (Interface.String*)null;
			{
				fixed (Interface.Callback* pointer = new Interface.Callback[1]) {
					*pointer = Converter.CreateCallback((delegate*<Interface.String*, Interface.StringList**, Interface.StringList**>)Marshal.GetFunctionPointerForDelegate(callbackProxy));
					callbackPointer = pointer;
				}
				fixed (Interface.String* pointer = new Interface.String[1]) {
					*pointer = Converter.CreateString(script);
					scriptPointer = pointer;
				}
				fixed (Interface.StringList* pointer = new Interface.StringList[1]) {
					*pointer = Converter.CreateStringList(argument);
					argumentPointer = pointer;
				}
			}
			var exceptionPointer = library.Execute(&callbackPointer, &scriptPointer, &argumentPointer, &resultPointer);
			var result = Converter.ParseString(ref *resultPointer);
			if (exceptionPointer != null) {
				var exception = Converter.ParseString(ref *exceptionPointer);
				throw new Exception(exception);
			}
			return result;
		}

		public static void Prepare (
			Library library
		) {
			{
			}
			var exceptionPointer = library.Prepare();
			if (exceptionPointer != null) {
				var exception = Converter.ParseString(ref *exceptionPointer);
				throw new Exception(exception);
			}
			return;
		}

		#endregion

	}

}
